### PR TITLE
remove reset_train_val_dataloaders(model) from Trainer and move data reloading logic to loop

### LINKED ISSUE #9502: Dataloader is reloaded twice after resuming from checkpoint
### Motivation

PyTorch Lightning reloads dataloader twice after resuming from checkpoint:
the first time reloads before train loop start [[link](https://github.com/PyTorchLightning/pytorch-lightning/blob/d2def369ec0ecdcf65acbe834c75cd2ff778d491/pytorch_lightning/trainer/trainer.py#L1122)];
The second reloads of default value for `reload_dataloaders_every_n_epochs` [[link](https://github.com/PyTorchLightning/pytorch-lightning/blob/6ff43cbff7a7b4ea75cd3d17b27aa7addf46421c/pytorch_lightning/loops/fit_loop.py#L202)]
This behavior could be problematic:
- This behavior could bring user confusion. They expect `reload_dataloaders_every_n_epochs = 1` forces the dataloaders to reload once every n epochs, but it loads multiple times sometimes. 
- The dataloader loading behavior is inconsistent between resuming from checkpoint and fresh training start. This could break the internal states of dataloader, which requires users to keep track of global states. 
- It may impact Fault tolerance training correctly captures the state of dataloader

### Pitch
Let's remove  `self.reset_train_val_dataloaders(model)` in `_run_train()` [[link](https://github.com/PyTorchLightning/pytorch-lightning/blob/d2def369ec0ecdcf65acbe834c75cd2ff778d491/pytorch_lightning/trainer/trainer.py#L1122)]
Because reset_*_dataloader() will be called in fit loop, there's no need to call it in trainer. 

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by ninginthecloud**
## What does this PR do?

Fixes #9502 

This PR aims to fix the bug mentioned in issue #9502. 
During fit loop, train_dataloader is loaded twice when resuming from checkpoint. This underlying behavior does not meet user's expectation. 
As discussed in an closed PR #9614, we will follow the second option to fix this issue: 
- **Option 2**: remove `self.reset_train_val_dataloaders(model)` in `Trainer._run_train()` and set the reload logics in corresponding loops. Additionally, 
1. `self.num_training_batches = 0` in `trainer._setup_on_init()`. This value is initialized as `self.num_training_batches = float('inf')`, 
2. `fit_loop.done` and `fit_loop.skip` has been updated. 
```python
@property
    def done(self) -> bool:
        """Evaluates when to leave the loop.

        Returns True if trainer.should_stop was set (e.g. by early stopping) or if the maximum number of steps or epochs
        is reached.
        """
        return stop_steps or should_stop or stop_epochs or self.trainer.num_training_batches == 0

@property
    def skip(self) -> bool:
        """Whether we should skip the training and immediately return from the call to :meth:`run`."""
        return self.done or self.trainer.limit_train_batches == 0
```
 
## Before submitting

- [X] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [X] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/base.py [hunk 1]
```diff
@@ -49,6 +49,7 @@ class Loop(ABC, Generic[T]):
 
     def __init__(self) -> None:
         self.restarting = False
+        self.is_fresh_start_epoch: bool = True
```
Comment (ananthsub) [inline pytorch_lightning/loops/base.py:52]: i do not think we should be introducing this attribute to the base loop interface. not all loops implement epoch-based semantics.

File: pytorch_lightning/loops/fit_loop.py [hunk 2]
```diff
@@ -161,12 +161,12 @@ def done(self) -> bool:
...
         self.trainer.should_stop = should_stop
 
-        return stop_steps or should_stop or stop_epochs
+        return stop_steps or should_stop or stop_epochs or self.trainer.num_training_batches == 0
 
     @property
     def skip(self) -> bool:
         """Whether we should skip the training and immediately return from the call to :meth:`run`."""
-        return self.done or self.trainer.num_training_batches == 0
+        return self.done or self.trainer.limit_train_batches == 0
```
Comment (ananthsub) [inline pytorch_lightning/loops/fit_loop.py:169]: @awaelchli @tchaton n00b question: why do we need `skip` at all vs directly calling `run` and having it exit?

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -511,7 +511,7 @@ def _setup_on_init(self, num_sanity_val_steps: int) -> None:
 
         self.should_stop = False
         self.state = TrainerState()
-        self.num_training_batches = 0
+        self.num_training_batches = float("inf")
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:514]: is this inconsistent if we don't do the same for validation/test/predict batches too?

File: tests/callbacks/test_progress_bar.py [hunk 4]
```diff
@@ -146,9 +146,10 @@ def test_progress_bar_fast_dev_run(tmpdir):
     assert 1 == progress_bar.val_batch_idx
     assert 0 == progress_bar.test_batch_idx
 
-    # the main progress bar should display 2 batches (1 train, 1 val)
-    assert 2 == progress_bar.main_progress_bar.total
-    assert 2 == progress_bar.main_progress_bar.n
+    # the main progress bar should display 1 batch, as max_steps = 1
+    # when fast_dev_run=True
+    assert 1 == progress_bar.main_progress_bar.total
+    assert 1 == progress_bar.main_progress_bar.n
```
Comment (ananthsub) [inline tests/callbacks/test_progress_bar.py:152]: this is indicating that this is a breaking change in the API

File: tests/trainer/test_data_loading.py [hunk 5]
```diff
@@ -115,8 +116,8 @@ def on_train_start(self, *_) -> None:
     else:
         warn_str = "Consider setting accelerator=ddp"
 
-    trainer = Trainer(accelerator="ddp_spawn")
     with pytest.warns(UserWarning, match=warn_str), pytest.raises(SystemExit):
+        trainer = Trainer(accelerator="ddp_spawn")
```
Comment (ananthsub) [inline tests/trainer/test_data_loading.py:120]: nit: what was wrong with initializing the trainer outside this context?

File: pytorch_lightning/loops/base.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:52]: Agree, this look like a hack and it won't generalize to all loops.

File: pytorch_lightning/loops/fit_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/fit_loop.py:169]: We believe it was cleaner to have a skip function if needed, so the code is slightly more organized. But it could be removed.

File: pytorch_lightning/loops/fit_loop.py [hunk 6]
```diff
@@ -188,7 +188,9 @@ def on_advance_start(self) -> None:
         model = self.trainer.lightning_module
 
         # reset train dataloader
-        if self.current_epoch != 0 and self.trainer._should_reload_dl_epoch:
```
Comment (tchaton) [inline pytorch_lightning/loops/fit_loop.py:191]: Hey @ninginthecloud. I might be wrong, but `self.current_epoch != 0` is just a small optimization to prevent re-creating the dataloader right ?

File: tests/callbacks/test_progress_bar.py (hunk 4 above)
Comment (tchaton) [reply tests/callbacks/test_progress_bar.py:152]: I believe this is wrong and the previous behaviour seemed accurate. the main_progress should get 2 batches: (1 train, 1 val)

File: tests/trainer/test_data_loading.py (hunk 5 above)
Comment (tchaton) [reply tests/trainer/test_data_loading.py:120]: I won't fail as the SystemExit is triggered within the `on_train_epoch_start ` of the Callback.

**Comment by carmocca** [comment]
Marking this as a draft as a merge seems to have gone wrong, mark it as ready when it's ready again!

**Comment by ninginthecloud** [comment]
> Marking this as a draft as a merge seems to have gone wrong, mark it as ready when it's ready again!

I've got rebase and made this PR in a clean stage. Looking for another round of review, thank you~ 😃 cc: @carmocca, @tchaton
